man/lsauxv.1
.\" VIEW THIS MAN PAGE WITH THE COMMAND:
.\" groff -mdoc -Tascii lsauxv.1
.Dd February 3, 2010
.Os Linux
.Dt LSAUXV \&1 "General Commands Manual"
.Sh NAME
.Nm lsauxv
.Nd command for querying the auxiliary vector
.Sh SYNOPSIS
.Nm lsauxv
.Op Fl Fl help | Fl Fl info | Fl Fl list | Fl Fl hwcap
.Nm lsauxv
.Ar a_type1 a_type2 ...
.Sh DESCRIPTION
.Pp
* WARNING: This library is going to be discontinued. Please use:
.Pa LD_SHOW_AUXV=1 /bin/true          
command for similar functionality instead.
.Pp 
This application takes one or more arguments, where each argument is a supported 
.Va a_type ,
and prints to standard out the corresponding 
.Va a_val 
value. If there are N 
.Va a_type 
arguements passed, then standard out will consists of N lines, where each line is the 
.Va a_val 
of the corresponding 
.Va a_type 
argument. Use 
.Op Fl Fl list 
to get a list of the supported 
.Va a_type 
arguments.
.Pp 
In this context an 
.Em a_type argument
refers to the 
.Va a_type 
member variable of the
.Vt auxv_t
structure, which is defined (on most systems) in 
.Pa /usr/include/elf.h .
For each 
.Va a_type 
argument, this application scans through the list of 
.Vt auxv_t 
structures in 
.Pa /proc/self/auxv 
until an 
.Vt auxv_t 
structure with a matching 
.Va a_type 
is found. The 
.Va a_val 
of the matching 
.Vt auxv_t 
structure is then printed out (one 
.Va a_val 
per line per 
.Va a_type 
parameter).
.Sh OPTIONS
.Bl -tag -width " "
.It Op Fl Fl help
Displays list of supported options and their descriptions.
.It Op Fl Fl info
Displays some basic information regarding the lsauxv application, as well as an example use.
.It Op Fl Fl list
Lists all supported a_type arguments.
.It Op Fl Fl hwcap
Lists the enabled hardware capabilities.
.It Ar a_type1 a_type2 ...
Replace 
.Ar a_type1 a_type2 ...
with one or more of the supported a_type arguments. The case of the arguments does not matter.
.El
.Sh FILES
.Bl -tag -width "/proc/self/auxv" -compact
.It Sy /proc/self/auxv
the file that this application scans when attempting to find an auxv_t structure with an a_type member that matches the argument(s).
.El
.Sh EXAMPLES
.Bl -item -compact
.It 
> lsauxv AT_PAGESZ AT_UID at_uid
.It 
4096
.It 
1016
.It 
1016
.El
.Sh DIAGNOSTICS
.Ex -std lsauxv
.Sh SEE ALSO
.Xr query_auxv 3
.Sh REPORTING BUGS
.Pp
Email bug reports to Raphael M. Zinsly <rzinsly@linux.vnet.ibm.com> or
Carlos Eduardo Seo <cseo@linux.vnet.ibm.com>.
.Sh AUTHORS
.Pp
Written by Sean Curry.


ChangeLog
2015-04-01 Rajalakshmi Srinivasaraghavan  <raji@linux.vnet.ibm.com>

	* config.log: Delete file.
	* config.status: Delete file.
	* .gitignore: New file.

2014-05-27  Adhemerval Zanella  <azanella@linux.vnet.ibm.com>

	* lsauxv.c (print_hwcap): Add support to HWCAP2.

	* Makefile.in [test_auxv_shared]: Fix link order to include library
	at the end.
	[test_auxv_fallback_shared]: Likewise.

2011-08-25  Ryan S. Arnold  <rsa@linux.vnet.ibm.com>

	* configure: Regenerated for libauxv 0.1.0.
	* config.h.in: Likewise.

2011-07-21  Tulio Magno Quites Machado Filho <tuliom@linux.vnet.ibm.com>

	* configure: Regenerated for version 0.1.0.
	* Makefile.in: Added static library build and testcase.
	* README: Added --enable-static.
	* configure.ac: Updated to version 0.1.0 and added RANLIB export.

2011-03-30  Ryan S. Arnold  <rsa@us.ibm.com>

	* man/auxv.3 (RATIONALE): Added new section with information from
	the README.

2011-03-30  Ryan S. Arnold  <rsa@us.ibm.com>

	* README: Updated sections on Dependencies, GNU/Linux OS, GLIBC
	Minimum Version, Configuration, Make Rules, and Use.
	* test_auxv.c (main): 'long' variables changed to 'unsigned long int'.
	Updated printf of hwcap to print leading zeros based on sizeof
	unsigned long int.
	* man/auxv.3 (EXAMPLES): Add example code which shows how to use all of
	the interfaces.  Add additional information on auxv/hwcap.h.
	(RETURN VALUES): Added return value description from prefetch_auxv().
	(SYNOPSIS): Added hwcap.h to synopsis.
	(DEPENDENCIES): Added information on libpthread dependency.
	(FILES): Added auxv/hwcap.h to the files list.

2011-03-30  Ryan S. Arnold  <rsa@us.ibm.com>

	* Makefile.in (install): Install lsauxv.1 into $(inst_mandir)/man1/
	and auxv.3 into man3/ rather than $(inst_mandir).  Create libauxv.3
	symlink to auxv.3.
	(uninstall): Remove $(inst_mandir)/man1/lsauxv.1 and
	$(inst_mandir)/man3/auxv.3.  Remove libauxv.3 symlink.


src/lsauxv.c
/* lsauxv - The Auxiliary Vector Query Application

   Author(s): Sean Curry <spcurry@us.ibm.com>
              Ryan S. Arnold <rsa@us.ibm.com>

   Copyright (c) 2010, IBM Corporation
   All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are met:
	 * Redistributions of source code must retain the above copyright
	   notice, this list of conditions and the following disclaimer.
	 * Redistributions in binary form must reproduce the above copyright
	   notice, this list of conditions and the following disclaimer in the
	   documentation and/or other materials provided with the distribution.
	 * Neither the name of the IBM Corporation nor the names of its
	   contributors may be used to endorse or promote products derived from
	   this software without specific prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED. IN NO EVENT SHALL IBM CORPORATION BE LIABLE FOR ANY
   DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
   (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
   SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
   CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
   OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
   DAMAGE.
 */

#include <stdio.h>
#include <strings.h>
#include <auxv/auxv.h>
#include <auxv/hwcap.h>

static void print_help(){
	printf("Usage: lsauxv [options]\n");
	printf("Options:\n");
	printf("--help\t\tDisplay this information\n");
	printf("--info\t\tDisplay some information about this application, as well as an example use.\n");
	printf("--list\t\tLists all supported a_type arguments.\n");
	printf("--hwcap\t\tPrint the system hardware capability (hwcap) information.\n");
	printf("[a_type]\tReplace [a_type] with one or more of the supported a_type's that are listed using --list.\n");
	printf("\t\tUse --info to find out more about this application, including what an a_type is.\n");
}

static void print_info(){
	printf("This application takes one or more arguments, where each argument is a supported a_type, and prints out\n");
	printf("the corresponding a_val. If there are N a_type arguments passed to lsauxv, then the standard output will\n");
	printf("consist of N lines, where each line is the a_val of the corresponding a_type argument. Use --list to get\n");
	printf("a list of the supported a_type arguments.\n\n");
	printf("In this context, an \"a_type\" argument refers to the a_type member of the auxv_t struct, which is defined\n");
	printf("in /usr/include/elf.h. For each a_type argument, the lsauxv application scans the auxv_t structures in\n");
	printf("/proc/self/auxv until a auxv_t structure with a matching a_type is found. The a_val of the matching auxv_t\n");
	printf("structure is then printed to standard out (one a_val per line per a_type parameter).\n\n");
	printf("Sample usage of lsauxv (with output): \n");
	printf("> lsauxv AT_PAGESZ AT_UID at_uid\n");
	printf("4096\n1016\n1016\n");
}

static void print_list(){
	printf("Below is a list of all supported arguments for the 'lsauxv' application. Use --help for more information about this application.\n\n");
	printf("Argument\tDescription\n");
	printf("--------\t-----------\n");
	printf("AT_NULL\t\tEnd of vector\n");
	printf("AT_IGNORE\tEntry should be ignored\n");
	printf("AT_EXECFD\tFile descriptor of program\n");
	printf("AT_PHDR\t\tProgram headers for program\n");
	printf("AT_PHENT\tSize of program header entry\n");
	printf("AT_PHNUM\tNumber of program headers\n");
	printf("AT_PAGESZ\tSystem page size\n");
	printf("AT_BASE\t\tBase address of interpreter\n");
	printf("AT_FLAGS\tFlags\n");
	printf("AT_ENTRY\tEntry point of program\n");
	printf("AT_NOTELF\tProgram is not ELF\n");
	printf("AT_UID\t\tReal uid\n");
	printf("AT_EUID\t\tEffective uid\n");
	printf("AT_GID\t\tReal gid\n");
	printf("AT_EGID\t\tEffective gid\n");
	printf("AT_CLKTCK\tFrequency of times()\n");
	printf("AT_PLATFORM\tString identifying platform\n");
	printf("AT_BASE_PLATFORM\tString identifying base platform\n");
	printf("AT_HWCAP\tMachine dependent hints about processor capabilities\n");
	printf("AT_FPUCW\tUsed FPU control word\n");
	printf("AT_DCACHEBSIZE\tData cache block size\n");
	printf("AT_ICACHEBSIZE\tInstruction cache block size\n");
	printf("AT_UCACHEBSIZE\tUnified cache block size\n");
	printf("AT_IGNOREPPC\tEntry should be ignored.\n");
	printf("AT_SECURE\tBoolean, was exec setuid-like?\n");
	printf("AT_SYSINFO\t\n");
	printf("AT_SYSINFO_EHDR\t\n");
	printf("AT_L1I_CACHESHAPE\t\n");
	printf("AT_L1D_CACHESHAPE\t\n");
	printf("AT_L2_CACHESHAPE\t\n");
	printf("AT_L3_CACHESHAPE\t\n");
}

static void print_hwcap(void)
{
	long hwcap_mask;
	long hwcap_mask2;
	hwcap_mask = query_auxv(AT_HWCAP);

	printf("HWCAP1 = 0x%08lX\n", hwcap_mask);
#ifdef __powerpc__
	if (hwcap_mask & PPC_FEATURE_32)
	    printf("  32-bit\n");

	if (hwcap_mask & PPC_FEATURE_64)
	    printf("  64-bit\n");

	if (hwcap_mask & PPC_FEATURE_601_INSTR)
	    printf("  HAS_601_INSTR\n");

	if (hwcap_mask & PPC_FEATURE_HAS_ALTIVEC)
	    printf("  HAS_ALTIVEC\n");

	if (hwcap_mask & PPC_FEATURE_HAS_FPU)
	    printf("  HAS_FPU\n");

	if (hwcap_mask & PPC_FEATURE_HAS_MMU)
	    printf("  HAS_MMU\n");

	if (hwcap_mask & PPC_FEATURE_HAS_4xxMAC)
	    printf("  HAS_4xxMAC\n");

	if (hwcap_mask & PPC_FEATURE_HAS_SPE)
	    printf("  HAS_SPE\n");

	if (hwcap_mask & PPC_FEATURE_HAS_EFP_SINGLE)
	    printf("  HAS_SPE_SINGLE\n");

	if (hwcap_mask & PPC_FEATURE_HAS_EFP_DOUBLE)
	    printf("  HAS_SPE_DOUBLE\n");

	if (hwcap_mask & PPC_FEATURE_NO_TB)
	    printf("  NO_TIMEBASE\n");

	if (hwcap_mask & PPC_FEATURE_POWER4)
	    printf("  POWER4 ISA\n");

	if (hwcap_mask & PPC_FEATURE_POWER5)
	    printf("  POWER5 ISA\n");

	if (hwcap_mask & PPC_FEATURE_POWER5_PLUS)
	    printf("  POWER5+ ISA\n");

	if (hwcap_mask & PPC_FEATURE_CELL_BE)
	    printf("  CELL BE\n");

	if (hwcap_mask & PPC_FEATURE_SMT)
	    printf("  HAS_SMT\n");

	if (hwcap_mask & PPC_FEATURE_ICACHE_SNOOP)
	    printf("  ICache Snoop\n");

	if (hwcap_mask & PPC_FEATURE_ARCH_2_05)
	    printf("  ARCH_2_05 ISA\n");

	if (hwcap_mask & PPC_FEATURE_ARCH_2_06)
	    printf("  ARCH_2_06 ISA\n");

	if (hwcap_mask & PPC_FEATURE_HAS_DFP)
	    printf("  HAS DFP\n");

	if (hwcap_mask & PPC_FEATURE_POWER6_EXT)
	    printf("  POWER6x instructions\n");

	if (hwcap_mask & PPC_FEATURE_HAS_VSX)
	    printf("  HAS VSX\n");
#endif

	hwcap_mask2 = query_auxv(AT_HWCAP2);
	printf("HWCAP2 = 0x%08lX\n", hwcap_mask2);
#ifdef __powerpc__
	if (hwcap_mask2 & PPC_FEATURE2_ARCH_2_07)
	    printf("  ISA 2.07\n");

	if (hwcap_mask2 & PPC_FEATURE2_HAS_HTM)
	    printf("  HTM\n");

	if (hwcap_mask2 & PPC_FEATURE2_HAS_DSCR)
	    printf("  HAS DSCR\n");

	if (hwcap_mask2 & PPC_FEATURE2_HAS_EBB)
	    printf("  HAS EBB\n");

	if (hwcap_mask2 & PPC_FEATURE2_HAS_ISEL)
	    printf("  HAS ISEL\n");

	if (hwcap_mask2 & PPC_FEATURE2_HAS_TAR)
	    printf("  HAS TAR\n");

	if (hwcap_mask2 & PPC_FEATURE2_HAS_VEC_CRYPTO)
	    printf("  HAS VEC CRYPTO\n");

	if (hwcap_mask2 & PPC_FEATURE2_HTM_NOSC)
	    printf("  HTM NOSC\n");
#endif
}

static ElfW(Addr) translate_a_type(char *str){
	if(!strcasecmp("AT_NULL", str)) return AT_NULL;
	if(!strcasecmp("AT_IGNORE", str)) return AT_IGNORE;
	if(!strcasecmp("AT_EXECFD", str)) return AT_EXECFD;
	if(!strcasecmp("AT_PHDR", str)) return AT_PHDR;
	if(!strcasecmp("AT_PHENT", str)) return AT_PHENT;
	if(!strcasecmp("AT_PHNUM", str)) return AT_PHNUM;
	if(!strcasecmp("AT_PAGESZ", str)) return AT_PAGESZ;
	if(!strcasecmp("AT_BASE", str)) return AT_BASE;
	if(!strcasecmp("AT_FLAGS", str)) return AT_FLAGS;
	if(!strcasecmp("AT_ENTRY", str)) return AT_ENTRY;
	if(!strcasecmp("AT_NOTELF", str)) return AT_NOTELF;
	if(!strcasecmp("AT_UID", str)) return AT_UID;
	if(!strcasecmp("AT_EUID", str)) return AT_EUID;
	if(!strcasecmp("AT_GID", str)) return AT_GID;
	if(!strcasecmp("AT_EGID", str)) return AT_EGID;
	if(!strcasecmp("AT_CLKTCK", str)) return AT_CLKTCK;
	if(!strcasecmp("AT_PLATFORM", str)) return AT_PLATFORM;
	if(!strcasecmp("AT_BASE_PLATFORM", str)) return AT_BASE_PLATFORM;
	if(!strcasecmp("AT_HWCAP", str)) return AT_HWCAP;
	if(!strcasecmp("AT_FPUCW", str)) return AT_FPUCW;
	if(!strcasecmp("AT_DCACHEBSIZE", str)) return AT_DCACHEBSIZE;
	if(!strcasecmp("AT_ICACHEBSIZE", str)) return AT_ICACHEBSIZE;
	if(!strcasecmp("AT_UCACHEBSIZE", str)) return AT_UCACHEBSIZE;
	if(!strcasecmp("AT_IGNOREPPC", str)) return AT_IGNOREPPC;
	if(!strcasecmp("AT_SECURE", str)) return AT_SECURE;
	if(!strcasecmp("AT_SYSINFO", str)) return AT_SYSINFO;
	if(!strcasecmp("AT_SYSINFO_EHDR", str)) return AT_SYSINFO_EHDR;
	if(!strcasecmp("AT_L1I_CACHESHAPE", str)) return AT_L1I_CACHESHAPE;
	if(!strcasecmp("AT_L1D_CACHESHAPE", str)) return AT_L1D_CACHESHAPE;
	if(!strcasecmp("AT_L2_CACHESHAPE", str)) return AT_L2_CACHESHAPE;
	if(!strcasecmp("AT_L3_CACHESHAPE", str)) return AT_L3_CACHESHAPE;

	return -1;
}

int main(int argc, char *argv[]){
	long i;
	ElfW(Addr) a_type;
	ElfW(Addr) a_val;
	
	if(argc == 1){
		printf("No arguments specified.\n\n");
		print_help();
		return 0;
	}
	
	for(i = 1; i < argc; i++){
		if(!strcasecmp("--help", argv[i]) || !strcasecmp("-help", argv[i])){
			print_help();
			return 0;
		}
		if(!strcasecmp("--info", argv[i]) || !strcasecmp("-info", argv[i])){
			print_info();
			return 0;
		}
		if(!strcasecmp("--list", argv[i]) || !strcasecmp("-list", argv[i])){
			print_list();
			return 0;
		}
		if(!strcasecmp("--hwcap", argv[i]) || !strcasecmp("-hwcap", argv[i])){
			print_hwcap();
			return 0;
		}

	}

	for(i = 1; i < argc; i++){
		a_type = translate_a_type(argv[i]);
		a_val = query_auxv(a_type);
#ifdef __powerpc64__
 		 printf("%lu\n", a_val);
#else
		printf("%u\n", a_val);
#endif
	}

	return 0;
}

